File: troco_enumeracao.rb
def troco_enumeracao(moedas, valor, quantidade_moedas=Hash.new(0))
  if valor == 0
    quantidade_moedas.each_pair do |moeda, quantidade|
      print "#{quantidade} moeda(s) de #{moeda}, "
    end
    print "\n"
  elsif !moedas.empty?
    moeda = moedas.first
    novas_moedas = moedas - [moedas.first]
    for i in 0..(valor/moeda)
      quantidade_moedas[moeda] += i
      troco_enumeracao(novas_moedas, valor-moeda*i, quantidade_moedas)
      quantidade_moedas[moeda] -= i
    end
  end
end

troco_enumeracao([3, 1], 10)

File: 2014.2-discretas/troco_enumeracao.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void trocoEnumeracaoRecursivo(int moedas[], int moedasSize, int valor, int *quantidadeMoedas, int moedasOriginalSize) {
  if (valor == 0) {
    for(int i = 0; i < moedasOriginalSize; i++) {
      if (quantidadeMoedas[i]) {
        printf("%d moeda(s) de %d, ", quantidadeMoedas[i], moedas[i]);
      }
    }
    printf("\n");
  } else if (moedasSize > 0) {
    int moeda = moedas[moedasSize-1];
    for(int i = 0; i <= valor/moeda; i++) {
      *(quantidadeMoedas+moedasSize-1) += i;
      trocoEnumeracaoRecursivo(moedas, moedasSize-1, valor-moeda*i, quantidadeMoedas, moedasOriginalSize);
      *(quantidadeMoedas+moedasSize-1) -= i;
    }
  }
}

void trocoEnumeracao(int moedas[], int moedasSize, int valor) {
  int *quantidadeMoedas = (int*) malloc(moedasSize*4);
  memset(quantidadeMoedas, 0, moedasSize*4);
  trocoEnumeracaoRecursivo(moedas, moedasSize, valor, quantidadeMoedas, moedasSize);
  free(quantidadeMoedas);
}

int main()
{
  int moedas[] = {9, 1};
  trocoEnumeracao(moedas, 2, 10);
  return 0;
}

